Bulldozer.CSV/Maps/Attributes.cs

```diff
@@ -127,165 +128,83 @@ private int LoadEntityAttributes( CSVInstance csvData )
-                        contextDbContext = Reflection.GetDbContextForEntityType( contextModelType );
-                        if ( contextDbContext != null )
-                        {
-                            contextService = Reflection.GetServiceForEntityType( contextModelType, contextDbContext );
-                        }
+                        contextService = Reflection.GetServiceForEntityType( contextModelType, contextDbContext );
                     }
 
-                    if ( entityTypeId.HasValue && contextService != null )
+                    if ( contextService != null )
                     {
-                        if ( !string.IsNullOrWhiteSpace( attributeIdString ) && !prevAttributeForeignKey.Equals( attributeForeignKey, StringComparison.OrdinalIgnoreCase ) )
-                        {
-                            attribute = attributeService.GetByEntityTypeId( entityTypeId ).FirstOrDefault( a => a.ForeignKey == attributeForeignKey );
-                            prevAttributeForeignKey = attributeForeignKey;
-                            prevRockKey = string.Empty;
-                        }
-                        else if ( string.IsNullOrWhiteSpace( attributeIdString ) )
-                        {
-                            // if no FK provided force attribute to null so the rockKey is tested
-                            attribute = null;
-                        }
-
-                        if ( attribute == null && !string.IsNullOrWhiteSpace( rockKey ) )
-                        {
-                            attribute = attributeService.GetByEntityTypeId( entityTypeId ).FirstOrDefault( a => a.Key == rockKey );
-                            if ( !prevRockKey.Equals( rockKey, StringComparison.OrdinalIgnoreCase ) )
-                            {
-                                prevRockKey = rockKey;
-                                prevAttributeForeignKey = string.Empty;
-                            }
-                        }
+                        MethodInfo qryMethod = contextService.GetType().GetMethod( "Queryable", new Type[] { } );
+                        var entityQry = qryMethod.Invoke( contextService, new object[] { } ) as IQueryable<IEntity>;
+                        var entityIdDict = entityQry.Where( e => e.ForeignKey != null && e.ForeignKey.StartsWith( this.ImportInstanceFKPrefix + "^" ) ).ToDictionary( k => k.ForeignKey, v => v );
+                        var attributeDict = attributeLookup.Where( a => a.EntityTypeId == entityTypeId ).ToDictionary( k => k.Key, v => v );
+                        var attributeDefinedValuesDict = attributeLookup.Where( a => a.EntityTypeId == entityTypeId && a.FieldTypeId == DefinedValueFieldTypeId ).ToDictionary( k => k.Key, v => definedTypeDict.GetValueOrNull( v.AttributeQualifiers.FirstOrDefault( aq => aq.Key == "definedtype" ).Value.AsIntegerOrNull().Value ).DefinedValues.ToDictionary( d => d.Value, d => d.Guid.ToString() ) );
```

Comment:
Can this use the common method?

Reply:
This is intentionally distinct. The initial db call to retrieve the full attribute list is at line 148 so that each iteration of the foreach loop is only manipulating the already fetched master list rather than more db calls